### PR TITLE
Feature/221 create revision on submit

### PR DESCRIPTION
**Pull request by todd-dembrey**

### CODE REVIEW

File: opentech/apply/funds/forms.py [hunk 1]
```diff
@@ -28,13 +28,8 @@ def clean_action(self):
         if not can_proceed(transition):
             action = self.instance.phase.transitions[action_name]
             raise forms.ValidationError(f'You do not have permission to "{ action }"')
-        self.transition = transition
```
Comment (zerolab) [inline opentech/apply/funds/forms.py]: Something to consider: I had to move progress_stage to the determination view: https://github.com/OpenTechFund/opentech.fund/pull/396/files#diff-69bf3b7f321569af3a40118621f7c5f3L82

File: opentech/apply/funds/views.py [hunk 2]
```diff
@@ -234,16 +224,16 @@ def get_form_class(self):
...
 
-        transition = set(self.request.POST.keys()) & set(self.transitions.keys())
+        action = set(self.request.POST.keys()) & set(self.transitions.keys())
 
-        if transition:
-            transition_object = self.transitions[transition.pop()]
-            self.object.get_transition(transition_object.target)(by=self.request.user)
-            self.object.save()
+        transition = self.transitions[action.pop()]
+        print(self.object.status)
```
Comment (zerolab) [inline opentech/apply/funds/views.py]: small note not to forget to clean up any stray `print`s

File: opentech/apply/funds/views.py [hunk 3]
```diff
@@ -79,15 +79,17 @@ def form_valid(self, form):
                 'apply:submissions:determinations:form',
                 args=(form.instance.id,)) + "?action=" + action)
 
+        form.instance.perform_transition(action, self.request.user)
+
         response = super().form_valid(form)
         return self.progress_stage(form.instance) or response
 
     def progress_stage(self, instance):
-        proposal_transition = instance.get_transition('draft_proposal')
```
Comment (zerolab) [inline opentech/apply/funds/views.py]: Something to consider: I had to move progress_stage to the determination view: https://github.com/OpenTechFund/opentech.fund/pull/396/files#diff-69bf3b7f321569af3a40118621f7c5f3L82

File: opentech/apply/funds/views.py [hunk 4]
```diff
@@ -211,19 +213,7 @@ def buttons(self):
     def get_form_kwargs(self):
         kwargs = super().get_form_kwargs()
         instance = kwargs.pop('instance')
-        form_data = instance.form_data
```
Comment (zerolab) [inline opentech/apply/funds/views.py]: nice cleanup!

**Comment by todd-dembrey** [comment]
Now includes 222 and 224. @chris-lawton can you check the styling changes please.

File: opentech/apply/funds/models.py [hunk 5]
```diff
@@ -714,26 +744,73 @@ def progress_application(self, **kwargs):
...
+        return self
+
+    def create_revision(self, draft=False, force=False, by=None, **kwargs):
+        self.clean_submission()
+        current_data = ApplicationSubmission.objects.get(id=self.id).form_data
+        if current_data != self.form_data or force:
+            if self.live_revision == self.draft_revision:
+                revision = ApplicationRevision.objects.create(submission=self, form_data=self.form_data, author=by)
+            else:
+                revision = self.draft_revision
```
Comment (zerolab) [inline opentech/apply/funds/models.py]: if you save draft, then go back make an edit and submit, the live revision become that of the draft, not of the submitted data. Will make test and fix

Comment (zerolab) [reply opentech/apply/funds/models.py]: sorted. this revision needed a save.

**Comment by zerolab** [comment]
Nice work!

**Pull request merged**